### ISSUE TITLE
Event-subscribed client segfaults if events cannot reach it

### ISSUE THREAD

**Issue by cpascual** [tango-controls/pytango#371]
I found this while investigating crashes in taurus GUIs when a tango DS was restarted (see  taurus-org/taurus#1118).

I managed to reproduce this using conda environments with both py2 and py3, and with pytango v 9.3.0 till 9.3.4-rc6 (pytango 9.2.x did not seem to be affected)

To reproduce, follow these steps  

(**UPDATE: 2020-07-09**: all this was done with the tangotest DS running on a separate docker bridge network from which the events could not reach the client running in the host)

1. Start tangotest DS
2. launch this script:
  ```python
  import tango
  import time
  import sys
  d = tango.DeviceProxy("sys/tg_test/1")
  cb = tango.utils.EventCallback()
  d.subscribe_event("ampli", tango.EventType.ATTR_CONF_EVENT, cb)
  while True:
      print(".")
      sys.stdout.flush()
      time.sleep(1)
  ```
3. stop and start the tangotest DS
4. ~10s after seeing the API_EventTimeout printed, the script crashes

Maybe it is related to #185 and/or #292

**Comment by cpascual** [comment]
Here is the backtrace when running it with gdb:

```
$ gdb --args python /home/cpascual/.PyCharmCE2019.3/config/scratches/scratch_93.py
(...)
(gdb) run
Starting program: /home/cpascual/miniconda/envs/kk3/bin/python /home/cpascual/.PyCharmCE2019.3/config/scratches/scratch_93.py
(...)
2020-07-08 17:53:47.276414 SYS/TG_TEST/1 AMPLI ATTR_CONF label='ampli'; unit=''
.
.
.
.
# TangoTest DS stopped and immediately started here
.
.
.
.
.
2020-07-08 17:53:59.446635 SYS/TG_TEST/1 AMPLI ATTR_CONF [API_EventTimeout] Event channel is not responding anymore, maybe the server or event system is down
.
.
.
.
.
.
.
.
.
.

Thread 3 "python" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffea381700 (LWP 27069)]
0x00007ffff6b8ef7e in Tango::EventConsumerKeepAliveThread::reconnect_to_zmq_channel(std::_Rb_tree_iterator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, Tango::channel_struct> >&, Tango::EventConsumer*, Tango::DeviceData&) () from /home/cpascual/miniconda/envs/kk3/lib/python3.7/site-packages/tango/../../../libtango.so.9
(gdb) bt
#0  0x00007ffff6b8ef7e in Tango::EventConsumerKeepAliveThread::reconnect_to_zmq_channel(std::_Rb_tree_iterator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, Tango::channel_struct> >&, Tango::EventConsumer*, Tango::DeviceData&) () from /home/cpascual/miniconda/envs/kk3/lib/python3.7/site-packages/tango/../../../libtango.so.9
#1  0x00007ffff6b91f84 in Tango::EventConsumerKeepAliveThread::main_reconnect(Tango::ZmqEventConsumer*, Tango::NotifdEventConsumer*, std::_Rb_tree_iterator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, Tango::event_callback> >&, std::_Rb_tree_iterator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, Tango::channel_struct> >&) ()
   from /home/cpascual/miniconda/envs/kk3/lib/python3.7/site-packages/tango/../../../libtango.so.9
#2  0x00007ffff6b9373e in Tango::EventConsumerKeepAliveThread::run_undetached(void*) () from /home/cpascual/miniconda/envs/kk3/lib/python3.7/site-packages/tango/../../../libtango.so.9
#3  0x00007ffff7f9fe48 in omni_thread_wrapper () from /home/cpascual/miniconda/envs/kk3/lib/python3.7/site-packages/tango/../../.././libomnithread.so.4
#4  0x00007ffff7f7dfa3 in start_thread (arg=<optimized out>) at pthread_create.c:486
#5  0x00007ffff7eae4cf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

```

**Comment by ajoubertza** [comment]
Hi @cpascual.

Thanks for the detailed example.  I'm struggling to reproduce the problem.  I've tried in various Docker containers, but don't get a segfault.  Is your TangoTest running in the same environment, so same version of cppTango?

Details of the versions tested after the example printout below.

```
(env-py3.7-tango9.3.4) root@8cfdee0e1bd4:/opt/project/tmp# python issue-371.py 
2020-07-08 20:00:13.756380 SYS/TG_TEST/1 AMPLI ATTR_CONF label='ampli'; unit=''
.
.
.
.
.
.
.
.
.
## Stopped TangoTest process about here
.
.
.
.
.
.
.
.
.
## Restarted TangoTest process about here
.
.
.
.
2020-07-08 20:00:35.744806 SYS/TG_TEST/1 AMPLI ATTR_CONF [API_EventTimeout] Event channel is not responding anymore, maybe the server or event system is down
2020-07-08 20:00:35.748993 SYS/TG_TEST/1 AMPLI ATTR_CONF label='ampli'; unit=''
.
.
.
.
.
.
.
.
## Stopped TangoTest process about here
.
.
2020-07-08 20:00:45.754437 SYS/TG_TEST/1 AMPLI ATTR_CONF [API_EventTimeout] Event channel is not responding anymore, maybe the server or event system is down
.
.
.
.
.
.
.
## Restarted TangoTest process about here
.
.
.
2020-07-08 20:00:55.764647 SYS/TG_TEST/1 AMPLI ATTR_CONF [API_EventTimeout] Event channel is not responding anymore, maybe the server or event system is down
2020-07-08 20:00:55.769823 SYS/TG_TEST/1 AMPLI ATTR_CONF label='ampli'; unit=''
.
.
.
.
.
.
.
.
.
.
.
.
``` 

Versions tested:
- PyTango 9.3.2 + cppTango 9.3.4rc6 + tango-test 3.0 (used for log shown above). In a miniconda container, using environment described [here](https://github.com/tango-controls/pytango/blob/develop/.devcontainer/environment-py3.7-tango9.3.4.yml).
- PyTango 9.3.1 + tango 9.3.2alpha1 (include cppTango and tango-test). In a miniconda container, using environment described [here](https://github.com/tango-controls/pytango/blob/develop/.devcontainer/environment-py3.7-tango9.3.2.yml).
- PyTango 9.3.2 + cppTango 9.3.4rc4 + TangoTest v?, using this Docker image:  nexus.engageska-portugal.pt/ska-docker/tango-pytango:9.3.2 (source [here](https://gitlab.com/ska-telescope/ska-docker/-/tree/7d5fc14a/docker/tango/tango-pytango) - the readme in that repo has details of the complicated hierarchy, but it is built from the TangoSourceDistribution 9.3.4rc4).
- PyTango 9.3.1 + cppTango 9.3.3  + TangoTest v?,, using this Docker image:  nexus.engageska-portugal.pt/ska-docker/tango-pytango:9.3.2 (source [here](https://gitlab.com/ska-telescope/ska-docker/-/tree/920258099/docker/tango/tango-pytango), built from the TangoSourceDistribution 9.3.3rc2).

TangoTest running in same container as client script.  Tango Database running a different container.

**Comment by cpascual** [comment]
Hi @ajoubertza , thanks for having a look at it!

I am running my server on a separate container, using SKA's tango images. More specifically, my tests were using 
`nexus.engageska-portugal.pt/ska-docker/tango-java:9.3.3` for the tangotest container, and `nexus.engageska-portugal.pt/ska-docker/tango-cpp:9.3.4-rc4` for the databaseds, and I started, stopped them using SKA's makefile (although I modified the tango.yml to map its port 10000 to localhost:10000)

Your comment actually made me think that this may be due to some communication problem (the docker bridge network may not allow some necessary communication with my client running in the host...).

I'll try to reproduce this without the docker infrastructure and come back with the results

**Comment by cpascual** [comment]
> this may be due to some communication problem (the docker bridge network may not allow some necessary communication with my client running in the host...).

It looks like that's it: I've tested running without docker containers (the DB running in the host, from the tango-db debian package) and the tangotestds running both from the tango-test_9.2.5a+dfsg1-2+b2 debian package and from conda (tango-test=3.0,  cpptango=9.3.4rc6). In both cases the issue is not reproduced.

Also I notice something that supports the hypothesis of event blocking cause: when the issue is reproduced, during the stop-start of the tangotestds I only see one config event being printed (the API_EventTimeout), but when it all goes fine, I see an additional config event (this can be seen when comparing https://github.com/tango-controls/pytango/issues/371#issuecomment-655607540 with https://github.com/tango-controls/pytango/issues/371#issuecomment-655736680 )

So, in conclusion, I think that this makes this issue less critical (since there are no crashes if the events can get through). But still I think it is worth checking it because tango clients should not crash even if events are blocked (e.g, by a firewall).

**Comment by cpascual** [comment]
I am updating the title of the issue according to the last findings
